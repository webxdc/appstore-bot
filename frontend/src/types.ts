import type { AppInfo } from './bindings/AppInfo'

export enum AppState {
  Initial,
  Downloading,
  DownloadCancelled,
  Received,
}

export interface AppInfoWithState extends AppInfo {
  state: AppState
}

export type AppInfosById = Record<string, AppInfoWithState>
