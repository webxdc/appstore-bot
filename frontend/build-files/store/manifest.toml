name = "Webxdc Store"
app_id = "store-xdc"
description = "store xdc"
source_code_url = "https://github.com/webxdc/store"
version = 2
