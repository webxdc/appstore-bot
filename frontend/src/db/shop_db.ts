import type { AppInfo } from '../bindings/AppInfo'
import type { AppInfoWithState } from '../types'
import type { XDCFile } from '../webxdc'

export class AppInfoDB {
  private dbName: string
  private db: IDBDatabase | undefined

  constructor(dbName: string) {
    this.dbName = dbName
  }

  open(): Promise<IDBDatabase> {
    if (this.db)
      return Promise.resolve(this.db)
    return new Promise((resolve, reject) => {
      const request = indexedDB.open(this.dbName, 3)
      request.onerror = () => reject(request.error)
      request.onsuccess = () => resolve((this.db = request.result))
      request.onupgradeneeded = () => {
        const db = request.result
        db.createObjectStore('appInfo', { keyPath: 'id' })
        db.createObjectStore('apps')
      }
    })
  }

  async insert(data: AppInfoWithState): Promise<void> {
    const db = await this.open()
    return new Promise((resolve, reject) => {
      const transaction = db.transaction('appInfo', 'readwrite')
      transaction.onerror = () => reject(transaction.error)
      const store = transaction.objectStore('appInfo')
      const request = store.add(data)
      request.onsuccess = () => resolve()
    })
  }

  async update(data: AppInfoWithState): Promise<void> {
    const db = await this.open()
    return new Promise((resolve, reject) => {
      const transaction = db.transaction('appInfo', 'readwrite')
      transaction.onerror = () => reject(transaction.error)
      const store = transaction.objectStore('appInfo')
      const request = store.put(data)
      request.onsuccess = () => resolve()
    })
  }

  async insertMultiple(data: AppInfoWithState[] | AppInfo[]): Promise<void> {
    const db = await this.open()
    return new Promise((resolve, reject) => {
      const transaction = db.transaction('appInfo', 'readwrite')
      transaction.onerror = () => reject(transaction.error)
      const store = transaction.objectStore('appInfo')
      data.forEach((item: AppInfoWithState | AppInfo) => store.add(item))
      transaction.oncomplete = () => resolve()
    })
  }

  async updateMultiple(data: AppInfoWithState[] | AppInfo[]): Promise<void> {
    const db = await this.open()
    return new Promise((resolve, reject) => {
      const transaction = db.transaction('appInfo', 'readwrite')
      transaction.onerror = () => reject(transaction.error)
      const store = transaction.objectStore('appInfo')
      data.forEach((item: AppInfoWithState | AppInfo) => store.put(item))
      transaction.oncomplete = () => resolve()
    })
  }

  async remove_multiple_app_infos(ids: number[]): Promise<void> {
    const db = await this.open()
    return new Promise((resolve, reject) => {
      const transaction = db.transaction('appInfo', 'readwrite')
      transaction.onerror = () => reject(transaction.error)
      const store = transaction.objectStore('appInfo')
      ids.forEach((id: number) => store.delete(id))
      transaction.oncomplete = () => resolve()
    })
  }

  async get_all(): Promise<AppInfoWithState[]> {
    const db = await this.open()
    return new Promise((resolve, reject) => {
      const transaction = db.transaction('appInfo', 'readonly')
      transaction.onerror = () => reject(transaction.error)
      const store = transaction.objectStore('appInfo')
      const request = store.openCursor()
      const result: AppInfoWithState[] = []
      request.onsuccess = () => {
        const cursor = request.result
        if (cursor) {
          result.push(cursor.value)
          cursor.continue()
        }
        else {
          resolve(result)
        }
      }
    })
  }

  // Add base64 encoded webxdc to the db.
  async add_webxdc(webxdc: XDCFile, id: number): Promise<void> {
    const db = await this.open()
    return new Promise((resolve, reject) => {
      const transaction = db.transaction('apps', 'readwrite')
      transaction.onerror = () => reject(transaction.error)
      const store = transaction.objectStore('apps')
      const request = store.add(webxdc, id)
      request.onsuccess = () => resolve()
    })
  }

  // Get base64 encoded webxdc from the db.
  async get_webxdc(id: number): Promise<XDCFile> {
    const db = await this.open()
    return new Promise((resolve, reject) => {
      const transaction = db.transaction('apps', 'readonly')
      transaction.onerror = () => reject(transaction.error)
      const store = transaction.objectStore('apps')
      const request = store.get(id)
      request.onsuccess = () => resolve(request.result)
    })
  }
}
