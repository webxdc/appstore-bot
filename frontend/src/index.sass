
.c-grid
  display: grid
  @apply items-center justify-stretch gap-2

.min-width
  width: 100%

@media screen and (min-width: 550px)
  .min-width
    width: 550px

  .c-grid
    @apply justify-center

input, textarea
  padding: 0.3rem !important
  @apply border rounded p-2

input:disabled
  @apply bg-gray-200 cursor-not-allowed

label
  @apply font-bold



@keyframes rotate
  to
    transform: rotate(-360deg)

.loading-spinner
  animation: rotate 1.5s linear infinite