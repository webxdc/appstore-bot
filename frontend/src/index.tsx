import { render } from 'solid-js/web'
import { createContext, createEffect, onMount } from 'solid-js'
import { createStore, unwrap } from 'solid-js/store'
import { Route, Router, Routes, hashIntegration } from '@solidjs/router'
import Info from './pages/info'
import Store from '~/pages/store'
import '~/index.sass'
import 'virtual:uno.css'
import '@unocss/reset/tailwind.css'
import type { StoreRequest } from '~/bindings/StoreRequest'

const metaData = {
  version: import.meta.env.VITE_COMMIT,
  updating: false,
  last_update: new Date(),
  last_update_serial: 0,
  // List of [app_id, version] for cached apps
  cached: [] as [string, number][],
}

const [meta, setMeta] = createStore(metaData)

const wrapper = {
  meta,
  setMeta,
  updateDone: (serial: number) => {
    setMeta('last_update', new Date())
    setMeta('last_update_serial', serial)
    setMeta('updating', false)
  },
  setCached: (cached: [string, number][]) => setMeta('cached', cached),
  update: () => {
    setMeta('updating', true)
    window.webxdc.sendUpdate({
      payload: { Update: { serial: meta.last_update_serial, apps: unwrap(meta.cached) } } as StoreRequest,
    }, '')
  },
}

export const metadataContext = createContext(wrapper)

render(
  () => {
    onMount(() => {
      const userString = localStorage.getItem('meta')
      if (!userString)
        return
      const parsed = JSON.parse(userString)
      setMeta(() => ({ ...parsed, last_update: new Date(parsed.last_update) }))
    })

    createEffect(() => localStorage.setItem('meta', JSON.stringify({ ...meta, last_update: meta.last_update.getTime() })))

    return (
      <metadataContext.Provider value={wrapper}>
        <Router source={hashIntegration()}>
          <Routes>
            <Route path="/" component={Store} />
            <Route path="/info" component={Info} />
          </Routes>
        </Router>
      </metadataContext.Provider>
    )
  },
  document.getElementById('root')!,
)
