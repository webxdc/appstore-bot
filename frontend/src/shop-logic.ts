import type { Setter } from 'solid-js'
import type { SetStoreFunction } from 'solid-js/store'
import { produce } from 'solid-js/store'
import type { AppInfoWithState, AppInfosById } from './types'
import { AppState } from './types'
import type { ShopResponse } from './bindings/ShopResponse'
import type { AppInfoDB } from './db/shop_db'
import { isOutdatedResponse, isUpdateSendResponse as isUpdateSentResponse } from './utils'

export type DownloadResponseOkay = Extract<ShopResponse, { type: 'DownloadOkay' }>
export type DownloadResponseError = Extract<ShopResponse, { type: 'DownloadError' }>
export type UpdateResponse = Extract<ShopResponse, { type: 'Update' }>

function isDownloadResponseOkay(p: any): p is DownloadResponseOkay {
  return p.type === 'DownloadOkay'
}

function isDownloadResponseError(p: any): p is DownloadResponseError {
  return p.type === 'DownloadError'
}

function isUpdateResponse(p: any): p is UpdateResponse {
  return p.type === 'Update'
}

function isEmpty(obj: any) {
  for (const prop in obj) {
    if (Object.prototype.hasOwnProperty.call(obj, prop))
      return false
  }
  return true
}

export function to_app_infos_by_id<T extends { app_id: string }>(app_infos: T[]): Record<string, T> {
  return app_infos.reduce((acc, appinfo) => {
    acc[appinfo.app_id] = appinfo
    return acc
  }, {} as Record<string, T>)
}

export async function updateHandler(
  payload: object,
  db: AppInfoDB,
  appInfo: AppInfosById,
  setAppInfo: SetStoreFunction<AppInfosById>,
  setlastUpdateSerial: Setter<number>,
  setIsUpdating: Setter<boolean>,
  setlastUpdate: Setter<Date>,
  setUpdateNeeded: Setter<boolean>,
  setUpdateReceived: Setter<boolean>,
) {
  if (isUpdateResponse(payload)) {
    if (isEmpty(appInfo)) {
      // initially write the newest update to state
      const app_infos = to_app_infos_by_id(payload.app_infos.map(app_info => ({ ...app_info, state: AppState.Initial } as AppInfoWithState)))
      setAppInfo(app_infos)
      await db.insertMultiple(Object.values(app_infos))
    }
    else {
      // all but the first update only overwrite existing properties
      const app_infos = to_app_infos_by_id(payload.app_infos.map((app_info) => {
        return { ...app_info, state: AppState.Initial }
      }))
      console.log('Reconceiling updates')
      const added: string[] = []
      const updated: string[] = []
      setAppInfo(produce((s) => {
        for (const key in app_infos) {
          if (s[key] === undefined) {
            s[key] = { ...app_infos[key] }
            added.push(key)
          }
          else {
            s[key] = Object.assign(s[key], { ...app_infos[key] })
            updated.push(key)
          }
        }
        for (const key of payload.updating) {
          s[key] = Object.assign(s[key], { state: AppState.Updating })
        }
      }))

      await db.insertMultiple(added.map(key => ({ ...app_infos[key], state: AppState.Initial })))
      await db.updateMultiple(updated.map(key => ({ ...app_infos[key], state: appInfo[key].state !== AppState.Initial ? AppState.Updating : AppState.Initial })))
    }

    setlastUpdateSerial(payload.serial)
    setIsUpdating(false)
    setlastUpdate(new Date())
  }
  else if (isDownloadResponseOkay(payload)) {
    const file = { base64: payload.data, name: `${payload.name}.xdc` }
    db.add_webxdc(file, payload.app_id)
    db.update({ ...appInfo[payload.app_id], state: AppState.Received })
    setAppInfo(payload.app_id, 'state', AppState.Received)
  }
  else if (isDownloadResponseError(payload)) {
    setAppInfo(payload.app_id, 'state', AppState.DownloadCancelled)
  }
  else if (isOutdatedResponse(payload)) {
    console.log('Current version is outdated')
    setUpdateNeeded(true)
  }
  else if (isUpdateSentResponse(payload)) {
    setUpdateReceived(true)
  }
}
