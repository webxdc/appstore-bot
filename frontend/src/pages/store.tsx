import type { Component } from 'solid-js'
import { Match, Show, Switch, createEffect, createMemo, createSignal, useContext } from 'solid-js'
import { For } from 'solid-js/web'
import { useStorage } from 'solidjs-use'
import Fuse from 'fuse.js'
import { createStore } from 'solid-js/store'
import { A } from '@solidjs/router'
import { metadataContext } from '~/index'
import type { StoreResponse } from '~/bindings/StoreResponse'
import type { StoreRequest } from '~/bindings/StoreRequest'
import { AppInfoDB } from '~/db/store_db'
import OutdatedView from '~/components/OutdatedView'
import { to_app_infos_by_id, updateHandler } from '~/store-logic'
import { AppState } from '~/types'
import type { AppInfoWithState, AppInfosById } from '~/types'
import mock from '~/mock'
import type { ReceivedStatusUpdate } from '~/webxdc'

const fuse_options = {
  keys: [
    'name',
    'author_name',
    'description',
  ],
  threshold: 0.4,
}

type DownloadResponseOkay = Extract<StoreResponse, { type: 'DownloadOkay' }>
type UpdateResponse = Extract<StoreResponse, { type: 'Update' }>

function AppInfoModal(item: AppInfoWithState, onDownload: () => void, onForward: () => void, onRemove: () => void) {
  const [isExpanded, setIsExpanded] = createSignal(false)

  return (
    <li class="w-full border rounded p-4 shadow">
      <div class="flex cursor-pointer items-center justify-between gap-2" onClick={() => setIsExpanded(!isExpanded())}>
        <img src={`data:image/png;base64,${item.image!}`} alt={item.name} class="h-20 w-20 rounded-xl object-cover" />
        <div class="flex-grow-1 overflow-hidden">
          <h2 class="text-xl font-semibold">{item.name}</h2>
          <p class="max-width-text truncate text-gray-600">{item.description}</p>
        </div>
        <Switch>
          <Match when={item.state === AppState.Initial}>
            <button class="send-button bg-blue-500" onClick={onDownload}>
              <div class="i-material-symbols:download text-white"></div>
            </button>
          </Match>
          <Match when={item.state === AppState.Downloading}>
            <div class="send-button bg-gray-500">
              <div class="i-eos-icons:bubble-loading text-white"></div>
            </div>
          </Match>
          <Match when={item.state === AppState.DownloadCancelled}>
            <div class="send-button bg-red-500" >
              <div class="i-material-symbols:error text-white"></div>
            </div>
          </Match>
          <Match when={item.state === AppState.Received}>
            <button class="send-button bg-green-500" onClick={onForward}>
              <div class="i-material-symbols:forward text-white"></div>
            </button>
          </Match>
          <Match when={item.state === AppState.Updating}>
            <div class="i-eos-icons:bubble-loading text-black"></div>
            <button class="send-button bg-green-500" onClick={onForward}>
              <div class="i-material-symbols:forward text-white"></div>
            </button>
          </Match>
        </Switch>
      </div >
      {
        isExpanded() && (
          <div class="flex flex-col">
            <p class="my-4 text-gray-600">{item.description}</p>
            <div class="my-2">
              <p class="text-sm text-gray-600"><span class="font-bold"> Submitter: </span>{item.submitter_uri}</p>
              <p class="break-all text-sm text-gray-600"><span class="font-bold"> Source code: </span>{item.source_code_url}</p>
              <p class="text-sm text-gray-600"><span class="font-bold"> Version: </span>{item.version}</p>
            </div>
            {(item.state === AppState.Received || item.state === AppState.Updating) && <button class="self-center btn-gray" onClick={onRemove}>Remove from cache</button>}
          </div>
        )
      }
      <div class="mt-1 flex justify-center" onClick={() => setIsExpanded(!isExpanded())}>
        <button class={`text-blue-800 ${isExpanded() ? 'i-carbon-up-to-top' : 'i-carbon-down-to-bottom'}`}>
        </button>
      </div>
    </li >
  )
}

interface AppListProps {
  items: AppInfoWithState[]
  search_query: string
  onDownload: (id: string) => void
  onForward: (id: string) => void
  onRemove: (id: string) => void
}

const AppList: Component<AppListProps> = (props) => {
  let fuse: Fuse<AppInfoWithState> = new Fuse(props.items, fuse_options)

  createEffect(() => {
    fuse = new Fuse(props.items, fuse_options)
  })

  const filtered_items = createMemo(() => {
    if (props.search_query !== '') {
      return fuse!.search(props.search_query).map(fr => fr.item)
    }
    else {
      return props.items
    }
  })

  return (
    <Show when={props.items.length !== 0} fallback={<p class="text-center unimportant">There are no apps</p>}>
      <For each={filtered_items() || props.items}>
        {
          item => AppInfoModal(item, () => props.onDownload(item.app_id), () => { props.onForward(item.app_id) }, () => props.onRemove(item.app_id))
        }
      </For>
    </Show>
  )
}

const Store: Component = () => {
  const [appInfo, setAppInfo] = createStore({} as AppInfosById)
  const [lastSerial, setlastSerial] = useStorage('last-serial', 0) // Last store-serial
  const [updateNeeded, setUpdateNeeded] = useStorage('update-needed', false) // Flag if the frontend is outdated
  const [updateReceived, setUpdateReceived] = useStorage('update-received', false)
  const [lastUpdateSerial, setlastUpdateSerial] = useStorage('last-update-serial', 0) // Last serial to initialize updateListener
  const [lastUpdate, setlastUpdate] = useStorage('last-update', new Date())

  const [query, setSearch] = createSignal('')
  const cached = createMemo(() => Object.values(appInfo).filter(app_info => app_info.state !== AppState.Initial))

  const { meta, setCached, update, updateDone } = useContext(metadataContext)
  createEffect(() => setCached(cached().map(app_info => ([app_info.app_id, app_info.version]))))

  // automatically update the app list
  const past_time = Math.abs(new Date().getTime() - lastUpdate().getTime()) / 1000
  if (appInfo === undefined || (past_time > 60 * 60)) {
    update()
  }

  const db = new AppInfoDB('webxdc')
  // This is for now _not_ synchronized with the update receival so a delayed
  // query could overwrite app updates. For now, this should be fine.
  db.get_all().then((apps) => {
    const app_infos = to_app_infos_by_id(apps)
    setAppInfo(app_infos)

    if (import.meta.env.DEV) {
      setAppInfo(to_app_infos_by_id(mock))
      setlastSerial(1)
    }
  })
  window.webxdc.setUpdateListener(async (resp: ReceivedStatusUpdate<UpdateResponse | DownloadResponseOkay>) => {
    updateHandler(resp.payload, db, appInfo, setAppInfo, updateDone, setUpdateNeeded, setUpdateReceived)
    setlastSerial(resp.serial)
  }, lastSerial())

  async function handleDownload(app_id: string) {
    setAppInfo(app_id, 'state', AppState.Downloading)
    window.webxdc.sendUpdate({
      payload: { Download: { app_id } } as StoreRequest,
    }, '')
  }

  async function handleForward(app_id: string) {
    const file = await db.get_webxdc(app_id)
    if (file === undefined) {
      throw new Error('No cached file found')
    }
    window.webxdc.sendToChat({ file })
  }

  async function handleRemove(app_id: string) {
    setAppInfo(app_id, 'state', AppState.Initial)
    db.remove_webxdc(app_id)
  }

  return (
    <OutdatedView critical={updateNeeded()} updated_received={updateReceived()}>
      <div class="c-grid p-3">
        <div class="min-width">

          {/* header */}
          <div class="flex items-center gap-2">
            <div>
              <h1 class="flex-shrink text-2xl font-bold">
                Webxdc Store
              </h1>
            </div>
            <A class="rounded-xl p-2 btn-gray" href='/info'>
              <Show when={meta.updating} fallback={
                <div class="border border-blue-500 rounded" i-carbon-information></div>
              }>
                <div class="loading-spinner border border-blue-500 rounded" i-material-symbols-sync></div>
              </Show>
            </A>
          </div>

          {/* app list */}
          <div class="mt-4 p-4">
            <ul class="w-full flex flex-col gap-2">
              <li class="w-full flex flex-col items-center justify-center gap-2">
                <div class="flex items-center justify-center gap-2">
                  <input class="border-2 rounded-2xl" onInput={event => setSearch((event.target as HTMLInputElement).value)} />
                  <button class="rounded-1/2 p-2 btn">
                    <div class="i-carbon-search" />
                  </button>
                </div>
              </li>
              <Show when={!(lastSerial() === 0)} fallback={
                <p class="text-center unimportant">Loading store..</p>
              }>
                <AppList
                  items={Object.values(appInfo)} search_query={query()}
                  onDownload={handleDownload}
                  onForward={handleForward}
                  onRemove={handleRemove} ></AppList>
              </Show>
            </ul>
          </div>
        </div >
      </div>
    </OutdatedView>
  )
}

export default Store
