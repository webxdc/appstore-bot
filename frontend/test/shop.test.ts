import { describe, expect, it, test, vi } from 'vitest'
import { type Setter, createRoot } from 'solid-js'
import type { SetStoreFunction } from 'solid-js/store'
import { createStore } from 'solid-js/store'
import { AppInfoDB } from '../src/db/shop_db'
import { AppState } from '../src/types'
import type { AppInfoWithState, AppInfosById } from '../src/types'
import type { DownloadResponseError, DownloadResponseOkay, UpdateResponse } from '../src/shop-logic'
import { to_app_infos_by_id, updateHandler } from '../src/shop-logic'
import type { WebxdcOutdatedResponse, WebxdcUpdateSentResponse } from '../src/utils'
import 'fake-indexeddb/auto'
import mock from '../src/mock'

const general_handlers = {
  appInfo: {} as AppInfosById,
  setAppInfo: ((() => { }) as SetStoreFunction<AppInfosById>),
  setlastUpdateSerial: ((() => { }) as Setter<number>),
  setIsUpdating: ((() => { }) as Setter<boolean>),
  setlastUpdate: ((() => { }) as Setter<Date>),
  setUpdateNeeded: ((() => { }) as Setter<boolean>),
  setUpdateReceived: ((() => { }) as Setter<boolean>),
}

describe('Shop receiving updates', () => {
  test('Handles outdated response', () => {
    const handlers = {
      db: new AppInfoDB('shoptesting'),
      ...general_handlers,
    }

    const payload = {
      type: 'Outdated',
      critical: true,
      version: 1,
    } as WebxdcOutdatedResponse

    const updateNeeded = vi.spyOn(handlers, 'setUpdateNeeded')
    updateHandler(payload, handlers.db, handlers.appInfo, handlers.setAppInfo, handlers.setlastUpdateSerial, handlers.setIsUpdating, handlers.setlastUpdate, handlers.setUpdateNeeded, handlers.setUpdateReceived)

    expect(updateNeeded).toHaveBeenCalledWith(true)
  })

  test('Handes update received', () => {
    const handlers = {
      db: new AppInfoDB('shoptesting'),
      ...general_handlers,
    }

    const payload = {
      type: 'UpdateSent',
    } as WebxdcUpdateSentResponse

    const setUpdateReceived = vi.spyOn(handlers, 'setUpdateReceived')
    updateHandler(payload, handlers.db, handlers.appInfo, handlers.setAppInfo, handlers.setlastUpdateSerial, handlers.setIsUpdating, handlers.setlastUpdate, handlers.setUpdateNeeded, handlers.setUpdateReceived)

    expect(setUpdateReceived).toHaveBeenCalledWith(true)
  })

  test('Handles download error', () => {
    const handlers = {
      db: new AppInfoDB('shoptesting'),
      ...general_handlers,
    }

    const payload = {
      type: 'DownloadError',
      error: 'error',
      app_id: 'test',
    } as DownloadResponseError

    const setAppInfo = vi.spyOn(handlers, 'setAppInfo')
    updateHandler(payload, handlers.db, handlers.appInfo, handlers.setAppInfo, handlers.setlastUpdateSerial, handlers.setIsUpdating, handlers.setlastUpdate, handlers.setUpdateNeeded, handlers.setUpdateReceived)
    expect(setAppInfo).toHaveBeenCalledWith(payload.app_id, 'state', AppState.DownloadCancelled)
  })

  test('Handles download okay', async () => {
    const db = new AppInfoDB('shoptesting')
    const handlers = {
      ...general_handlers,
      db,
      appInfo: to_app_infos_by_id(mock),
    }

    await db.insertMultiple(mock)

    const payload = {
      type: 'DownloadOkay',
      app_id: mock[0].app_id,
      name: 'test',
      data: 'test',
    } as DownloadResponseOkay

    const setAppInfo = vi.spyOn(handlers, 'setAppInfo')
    updateHandler(payload, handlers.db, handlers.appInfo, handlers.setAppInfo, handlers.setlastUpdateSerial, handlers.setIsUpdating, handlers.setlastUpdate, handlers.setUpdateNeeded, handlers.setUpdateReceived)
    expect(await db.get_webxdc(payload.app_id)).matchSnapshot()
    expect(await db.get(payload.app_id)).toStrictEqual({ ...mock[0], state: AppState.Received })
    expect(setAppInfo).toHaveBeenCalledWith(payload.app_id, 'state', AppState.Received)
  })

  test('Handles new AppIndex', async () => {
    const db = new AppInfoDB('shoptesting2')
    const handlers = {
      db,
      ...general_handlers,
      appInfo: {},
    }

    const payload = {
      type: 'Update',
      app_infos: mock,
      serial: 12,
    } as UpdateResponse

    const setAppInfo = vi.spyOn(handlers, 'setAppInfo')
    await updateHandler(payload, handlers.db, handlers.appInfo, handlers.setAppInfo, handlers.setlastUpdateSerial, handlers.setIsUpdating, handlers.setlastUpdate, handlers.setUpdateNeeded, handlers.setUpdateReceived)

    const initial_mock = mock.map(app_info => ({ ...app_info, state: AppState.Initial, cached: false } as AppInfoWithState))
    expect(setAppInfo).toHaveBeenCalledWith(to_app_infos_by_id(initial_mock))
    expect(await db.get_all()).toStrictEqual(Object.values(initial_mock))
  })

})
