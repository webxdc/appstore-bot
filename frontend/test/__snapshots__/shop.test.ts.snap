// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shop receiving updates > Handles download okay 1`] = `
{
  "base64": "test",
  "name": "test.xdc",
}
`;
